### ISSUE TITLE
OpenID Connect

### ISSUE THREAD

**Issue by francescocarzaniga** [photoprism/photoprism#782]
It seems that multi-user instances are on your radar, so I would like to suggest implementing OIDC for the login flow. OIDC is becoming the de-facto standard and there are tons and tons of projects (Open Source and not) using it now. It could be used together with [LDAP](https://github.com/photoprism/photoprism/issues/484) using an IdP like Keycloak, so it would cover all bases there and not require you to implement other Directory services.
I would personally suggest looking at [Owncloud](https://github.com/owncloud/ocis) for inspiration (also for the Open Source/commercial side).

**Comment by bnounours** [comment]
+1
For now either the collection is public (which is not ok for personal photos) either the admin account need to be connected (which is not ok to give admin account to all users)

**Comment by butonic** [comment]
For OCIS we integrated konnectd from kopano as an OpenID Identity Provider. It can use an LDAP server to provide users. A minimal version could use glauth as the user backend. Personally, I'd love to see a reusable minimal set of services to provide users for open source projects that want to embrace OpenID Connect.  It sucks to implement user authentication. Small deployments don't need a full blown keycloak server but should be able to embed a small scale solution. We did it with ownCloud and we did it again for ocis. It should be possible to flesh out something that can be reused by all. Konnectd and glauth are written in go and with ocis we tried to go for a microservice architecture. Maybe we could add Photoprism as one of the services? Happy to discuss ways forward. Ping me on https://talk.owncloud.com. 

Cheers and Happy New year!

**Comment by Nicarim** [comment]
Hey - got a question regarding this issue. Is this something you're considering implementing to the project? I might want to contribute the code to repository but I'd like to know how would you prefer to do it.

My initial idea (and reason) for adding SSO (or OpenID, or whatever) was to initially support reverse proxy functionality authentication middle like for example in Traefik authentication middlewares.

Good example of such configuration can be [found here](https://www.authelia.com/docs/deployment/supported-proxies/traefik2.x.html) for authelia SSO. 

I'd like to gate access to my private cloud through SSO solution like Authelia. They provide a way to inject HTTP header called `Remote-User` or whatever name you give it - so that application receiving request can trust that header and that user is authorized and their name is what is in that header. Example implementation of it [with Jira](https://www.authelia.com/docs/community/using-remote-user-header-for-sso-with-jira.html) out of their documentation.

Getting to the point - would you accept contribution to get `Remote-User` header be a way to authorize the user. 

Perhaps initially could that be whatever is in that header is treated as `admin` user, but if there is no such header then you're prompted to log in?

Example Flow Unauthorized:
- User enters `example.com/photoprism`
- Traefik sees user is not authenticated by asking Authelia if their credentials in cookies / session are correct
- User is redirected to independent flow of authentication provided by Authelia
- Authelia completes authentication, redirects back to `example.com/photoprism` and injects header `Remote-User: youruserfromSSO` 
- PhotoPrism sees that header and automatically authenticates user as `admin`

Of course assumption is that this header is dropped if provided from the internet and it is only injected by trusted middleware.

Just to be sure - You can achieve this functionality right now by setting the photoprism instance to **public**, but that removes authentication for other endpoints like WebDAV endpoint which makes it impossible to have public instance with authorization on that endpoint. I had an idea of making exception for WebDAV endpoint that it would accept PhotoPrism native credentials, and that visiting through browser normally would use SSO. For that to happen - above idea needs to be somehow implemented.

Tell me your thoughts about it :)

Not sure if I should post it under OpenID or separate issue, so if you want me to create new issue I can do that.

**Comment by lastzero** [comment]
Already working on it.

**Comment by Nicarim** [comment]
Full OpenID authenication or just the trusted header option?

**Comment by moximoti** [comment]
Hi, We’re currently implementing user management and OpenID Connect Client (RP) in PhotoPrism. This does not include trusted headers. I suggest to wait until user management and OIDC are properly implemented before aiming for a third authentication method, as we need to integrate all this with our current session management.

Btw, Authelia also supports OIDC. So there’s actually no need to implement trusted headers for your use case once we have OIDC ;)

**Comment by Nicarim** [comment]
That's correct - OIDC will work just as well if there is OIDC. 

I just thought that if you wouldn't want to waste too much time on that trusted headers could be a good compromise - after all it would be just some IF somewhere granting unconditional access instead of checking password.

Just a question then - what about the WebDAV scenario where I still want to be able to use PhotoPrism native credentials? Or do you have some other solution to that problem? if OIDC will be used then the application would need to be able to get the authentication somehow through OIDC - and none of the apps like photosync would support that I guess.

**Comment by lastzero** [comment]
We should provide app passwords / tokens for that.

**Comment by burnoutberni** [comment]
Any updates on when to expect OIDC support to land? I've noticed that there's been some additional work going on in the [`feature/oidc-v2`](https://github.com/photoprism/photoprism/tree/feature/oidc-v2) branch, but these changes haven't been referenced here – are they connected?

**Comment by lastzero** [comment]
Hopefully in a few weeks. We do our best, but are limited by the constraints of physics. Check out the list of changes in our upcoming release: https://docs.photoprism.org/release-notes/

**Comment by mescanne** [comment]
This is one feature I'm quite excited for. Everything I have is behind pomerium (OpenID-protected reverse proxy), so this would fit in very nicely.

**Comment by burnoutberni** [comment]
I'll keep my eyes peeled, thanks for the update.

**Comment by tomlawesome** [comment]
Really keen for this, I've deployed PhotoPrism and had a play. It looks fantastic but without SSO support, I find apps/services I add don't get used very much as users don't want to register. If they can click through and login with the same username and have an account autoprovisioned it's a whole different story! 

Hoping I can ditch Synology photos when this/multi-user is live!

**Comment by lastzero** [comment]
As you can see in our [Release Notes](https://docs.photoprism.app/release-notes/) and on the public [Roadmap](https://github.com/photoprism/photoprism/projects/5), we've decided not to release additional multi-user features before Christmas:

- This is partly because unexpected [security vulnerabilities](https://docs.photoprism.app/release-notes/) in Go and Log4j surfaced this month and we have a zero bug policy.
- In addition to providing support and fixing bugs discovered in the stable release, multi-user support is currently a top priority.
- Features are not released until we deem them safe and bug free. This is in the best interest of all users, part of our culture, and can't be debated.
- Users can do their part and [become a sponsor](https://docs.photoprism.app/funding/) to get their favorite features as soon as possible. At least until the funding goal is reached, please do not ask for deadlines.

**Comment by Akruidenberg** [comment]
Is OpenID available for testing in the preview build? Would like to test this feature with Authentik

**Comment by lastzero** [comment]
1. We [received a lot of support requests and other questions](https://www.reddit.com/r/photoprism/comments/s26o55/since_we_receive_many_more_support_requests_than/) over the holidays and in the last few days. More than our team can handle, and there have been very few community members who jumped in.
2. As a result, our priority was to expand the [Troubleshooting Guide](https://docs.photoprism.app/getting-started/troubleshooting/) and divide it into subsections to make it easier for users to find common issues and fix them on their own. Many other parts of our docs were improved as well, for example [Connecting via WebDAV](https://docs.photoprism.app/user-guide/sync/webdav/).
3. We felt that this made more sense than adding more and more features as quickly as possible.

**Comment by rpatel3001** [comment]
I just want to chime in and +1 trusted headers. All my services are behind traefik with traefik-forward-auth and use the X-Forwarded-User header.

**Comment by dada513** [comment]
Any progress?

**Comment by graciousgrey** [comment]
We have a zero bug policy and do our best to help users when they need support or have other questions. 

This comes at a price, as we can't give exact deadlines for new features. Also keep in mind that epics like multi-user support can't be delivered in a single update. Working on big features is a continuous process that can take months to years until all use cases are covered.

Having said that, funding really has the highest impact. So users can do their part and [become a sponsor](https://photoprism.app/membership) to get their favorite features as soon as possible.

**Comment by ghost** [comment]
This feature is a must-have for anything I expose to the public internet. 👍 👍

**Comment by especially-relative** [comment]
Being to manage family accounts in keycloak, rather than on a per-app basis
is huge for me, I am following this, and when progress is made (or if
testers are needed ;)) I'll give it a spin!

**Comment by JeanCGR** [comment]
Regarding OpenID authentication, when will the feature be added in the main branch ?
Can we use OpenID authentication currently ?
Thank you for your answers guys

**Comment by lastzero** [comment]
We have a zero-bug policy and are currently working on the next stable release, which we don't want to delay any further. Our priorities are generally as follows:

https://www.photoprism.app/oss/faq#priorities

If you need this for your business and purchase a commercial license with support, we can assign a higher priority to this issue.

**Comment by lastzero** [comment]
Since people now seem to be [starting to create forks](https://github.com/photoprism/photoprism/pull/995#issuecomment-1517397788) of our software to [monetize it](https://github.com/photoprism/photoprism/pull/995#issuecomment-1519017608), we will be implementing OpenID Connect support as part of PhotoPrism Plus. The code will then be in a separate repo, but our official builds and Docker images can be downloaded and used by anyone, including app store providers. You also don't need a membership to use OpenID Connect. However, our Plus license includes limitations to ensure that our work is not unfairly monetized by entrepreneurs, SaaS providers, and venture capital-backed startups:

https://www.photoprism.app/plus/license

We think this is a pragmatic compromise that ensures that most of our code remains open source and much of the functionality remains available for free. In a perfect world, of course, it would not be necessary to use different licenses (or even the AGPL instead of a permissive license like BSD or MIT).

**Comment by svengreb** [comment]
It's sad that this is required these days. I love OSS, but I also see that such steps are required to protect projects from "parasitic use" (looking at #995). I also don't see how this feature is something that is required for home users anyway so this is a good and very understandable candidate for being part of the commercial editions.

**Comment by mateuszdrab** [comment]
> It's sad that this is required these days. I love OSS, but I also see that such steps are required to protect projects from "parasitic use" (looking at #995). I also don't see how this feature is something that is required for home users anyway so this is a good and very understandable candidate for being part of the commercial editions.

Just a quick 2 cents from me, the use case I had for OIDC in Photoprism (and Home Assistant) is that I maintain a homelab and rely on Azure AD for SSO authentication. 

This allows people (parents) who are not technical to use things that require authentication in a secure, passwordless way as the device is registered and uses device authentication. 

OIDC does sound like a commercial feature; however, it can have a good use at home in big family setups where it's too much work to maintain multiple users and their accounts. Perhaps OIDC could make it's way into the free tier, with a user count limitation or into the essentials tier with the same limitation. 

If I had to pay for Photoprism as is now, I'd not be able to. I am not satisfied with the performance of the object categorization feature (tried to make an improvement there in my PR but this will need further discussion and rearchitecting potentially for tflite) and the other arguments are that there are no native Android apps so I never ended up adopting Photoprism into as much use as I wanted.

If I was able to put a bounty on a feature, I'd definitely pay for OIDC and improvement of the AI features because Photoprism is a really cool solution.

**Comment by lastzero** [comment]
> Just a quick 2 cents from me, the use case I had for OIDC in Photoprism (and Home Assistant) is that I maintain a homelab and rely on Azure AD for SSO authentication.

We understand that this is a dual-use technology, just as some enthusiasts run complete enterprise stacks at home. However, the majority of home users probably won't need it, while all businesses will need either OpenID Connect or LDAP/AD.

> Perhaps OIDC could make it's way into the free tier, with a user count limitation or into the essentials tier with the same limitation.

As mentioned in my comment above, you don't need a membership to use OpenID Connect. However, we will not include the code in the main repository under the AGPL license. If you want to distribute this code or software based on it, you will need to ask for permission, which we don't think should cause any problems for private users.

> If I had to pay for Photoprism as is now, I'd not be able to. I am not satisfied with the performance [...]

If you want us to develop more features or improve the performance, it seems like a good idea to support us more, not less? The Essentials tier is less than 2 Euros per month. That shouldn't stop anyone from signing up.

**Comment by girlpunk** [comment]
> Since people now seem to be [starting to create forks](https://github.com/photoprism/photoprism/pull/995#issuecomment-1517397788) of our software to [monetize it](https://github.com/photoprism/photoprism/pull/995#issuecomment-1519017608), we will be implementing OpenID Connect support as part of PhotoPrism Plus. 

With all due respect, if the PhotoPrism developers did not want their project used for commercial purposes, why was a licence that allows this chosen?

> If you want us to develop more features or improve the performance, it seems like a good idea to support us more, not less?

Therein lies the difference between asking people to pay for a product, and to pay for further development of said product.

**Comment by grachevko** [comment]
It's hard for me to imagine that a company would consider SSO as the primary reason for purchasing a license.
In most cases, a business needs technical support, because stable work is cost a lot of money.

*another one ~~enthusiasts~~ developer with complete enterprise stacks at home k8s cluster

**Comment by lastzero** [comment]
> With all due respect [...]

As mentioned in our [Code of Conduct](https://www.photoprism.app/code-of-conduct), we kindly ask that you do not use GitHub Issues or other development tools to start general discussions or express personal opinions.

> If the PhotoPrism developers did not want their project used for commercial purposes, why was a licence that allows this chosen?

We stand by our license and do not "forbid" something that was previously allowed. In fact, I have said from the beginning that this will not be a free enterprise solution:

- https://github.com/photoprism/photoprism/issues/98#issuecomment-569683052

This is why we use a different license for certain upcoming features. The entitlement of some users and the way people talk to us now confirm this decision.

**Comment by lastzero** [comment]
> It's hard for me to imagine that a company would consider SSO as the primary reason for purchasing a license.

If you read carefully, you'll see that my original comment wasn't about licensing to potential business customers, but about an entrepreneur who publicly announced that he was considering working against us so that he could better monetize our software with the least amount of work.

> *another one enthusiasts developer with complete enterprise stacks at home k8s cluster

Again, the OpenID Connect feature will be free, like in free beer.

**Comment by AllenBalaj** [comment]
> We understand that this is a dual-use technology, just as some enthusiasts run complete enterprise stacks at home. However, the majority of home users probably won't need it, while all businesses will need either OpenID Connect or LDAP/AD.

Ultimately, the PhotoPrism team can do whatever they want to, because they're the ones writing and charging for the software. That being said, as someone who's been writing software professionally for several decades, I don't understand how other developers are still under the impression that SSO is something that most home users won't need. Who's running a bunch of services at home for friends or family and has the time to set up the same users for 20-30+ different sets of software? I'd genuinely be interested in hearing from home users that have more than a handful of services and  don't require SSO these days. If anything, I would think that home users need SSO more than an business would, because it's usually one person maintaining the setup and they have less time and money to do it in. Personally, I don't even bother adding new services to my K8S cluster unless I can easily add/remove users in a secure way, especially internet-facing services. https://sso.tax/ was created to explain this in a succinct way.

What I've seen across the projects we've adopted internally (both for family use and within my own businesses) is that orgs that want to charge either implement a user limit, license advanced functionality, sell support, or offer paid hosting. It sounds like that's the direction PhotoPrism is taking here, which is great, I just wish that there were more consensus and understanding of the fact that SSO is a basic modern security feature similar to having an API and supporting containerized implementations. If a FOSS project doesn't offer that then we should be contributing to help make it happen or migrating to other projects if that functionality won't be prioritized.

**Comment by lastzero** [comment]
Can we please move this discussion to GitHub Discussions? 🙄

**Comment by lastzero** [comment]
As you can see from the priority label, this feature has priority and will also be free. Note that abusing GitHub Issues for general discussion distracts us from getting the next stable version ready for release, which is why comments are closed now.

**Comment by lastzero** [comment]
See issue #808 for an update regarding the implementation of 2FA, OpenID Connect and Passkey support:
- https://github.com/photoprism/photoprism/issues/808#issuecomment-1828532810

**Comment by jerrac** [comment]
I was just reevaluating PhotoPrism for sharing family pictures between family members. I like a lot of what I see, and might go ahead and use it. But I'd really like to have some kind of SSO available, because I have a Keycloak instance that I use to help make managing accounts easier.. Which is why I found this issue. :) 

I'm commenting because I want to mention that I hope you don't lock the feature up in your paid version like the account UI is locked. 

As an avid believer in self-hosting, it's been painful to see how many open source projects lock SSO support up as "enterprise" when SSO is really not an enterprise level tool anymore. Maybe 5 or 10 years ago, but in 2024, SSO (and MFA) are essential to any level of webapp.

**Comment by svengreb** [comment]
> As an avid believer in self-hosting […]

This is a double-edged sword: You believe in self-hosting, including OSS, but want all of it for free without contributing or support the creators in any way? This thinking is the reason why problems like the _xz_ backdoor happen and OSS dies step-by-step. The PhotoPrism team has gone the, kind of risky, way of trying to make a living from an OSS project while also still providing so many features for free. Everyone could simply go ahead and implement it on their own since the code is available, but instead everyone waits until the spoon reaches the mouth without any effort on their own.

> Maybe 5 or 10 years ago, but in 2024, SSO (and MFA) are essential to any level of webapp […]

Let's take this into the real world… why don't you ask the cashier to give you the groceries for free since production can be easily automated these days with all of the modern machinery. While we're at it, let's also get a new car since fully automated robots in production lines also replaced many humans so a cars are also "essentials to any level of industrialized countries".

It's sad that nowadays you always have to make it clear to people that they don't even have the slightest right to make demands on free resources.

I hope you see where this goes and my comment is extra sarcastic since this exploiting behavior of "I have demands for everything I get for free" has been successfully placed by large companies into the minds of the younger generation and as a maintainer this keeps sucking the motivating energy out of anyone who enjoys creating something.

**Comment by tomlawesome** [comment]
Yawn. Your examples are deliberately and excessively dramatic. 

No one’s forcing anyone to provide anything for free. If the devs don’t want to implement it they won’t, and if you’re saying that SSO is the only feature they have to drive income then I think you’re firstly doing PhotoPrism a discredit and secondly just pointing out a lack of commercial viability. 




> > As an avid believer in self-hosting […]
> 
> This is a double-edged sword: You believe in self-hosting, including OSS, but want all of it for free without contributing or support the creators in any way? This thinking is the reason why problems like the _xz_ backdoor happen and OSS dies step-by-step. The PhotoPrism team has gone the, kind of risky, way of trying to make a living from an OSS project while also still providing so many features for free. Everyone could simply go ahead and implement it on their own since the code is available, but instead everyone waits until the spoon reaches the mouth without any effort on their own.
> 
> > Maybe 5 or 10 years ago, but in 2024, SSO (and MFA) are essential to any level of webapp […]
> 
> Let's take this into the real world… why don't you ask the cashier to give you the groceries for free since production can be easily automated these days with all of the modern machinery. While we're at it, let's also get a new car since fully automated robots in production lines also replaced many humans so a cars are also "essentials to any level of industrialized countries".
> 
> It's sad that nowadays you always have to make it clear to people that they don't even have the slightest right to make demands on free resources.
> 
> I hope you see where this goes and my comment is extra sarcastic since this exploiting behavior of "I have demands for everything I get for free" has been successfully placed by large companies into the minds of the younger generation and as a maintainer this keeps sucking the motivating energy out of anyone who enjoys creating something.

**Comment by lastzero** [comment]
Here's some context to help you understand what @svengreb was referring to, in case you haven't heard about it (yet):
- https://robmensching.com/blog/posts/2024/03/30/a-microcosm-of-the-interactions-in-open-source-projects/

We plan to make OIDC part of our Community Edition, and have communicated this publicly on several occasions, for example:
- https://github.com/photoprism/photoprism/issues/808#issuecomment-1828532810

At the same time, it's true that demands for free features (or for us to work faster) usually come from people we've never heard of and who [don't actively contribute to the project](https://www.photoprism.app/oss/faq) by [either helping out](https://docs.photoprism.app/developer-guide/) or [becoming a member](https://www.photoprism.app/editions#compare). I also have to agree that this kind of communication sucks the motivation out of you as a maintainer, especially when you put in so much work mostly for free.

On a side note, it would be great if end users did not use this issue tracker for discussions. Instead, you are welcome to use [GitHub Discussions](https://github.com/photoprism/photoprism/discussions) for that, as mentioned in our [README.md](https://github.com/photoprism/photoprism?tab=readme-ov-file#github-issues-%EF%B8%8F) file and throughout the documentation:
- https://github.com/photoprism/photoprism/discussions

**Comment by jerrac** [comment]
> We plan to make OIDC part of our Community Edition ...

My searches on the topic did not turn up the info about making the feature part of the community edition. I'm glad to hear that you are doing so. :D Thanks!

> On a side note, it would be great if end users did not use this issue tracker for discussions...

I commented here on the issue because my comment is directly relevant to how the feature is implemented. So, in this case, I thought the it was fine for the issue queue and didn't go looking very hard elsewhere. 

Side node: I will say that one of the more frustrating things about Open Source is how many different places you can go to communicate. I've spent a significant amount of time over the years just trying to make sure I post in the correct place. It is very frustrating to do your best and then be told off for putting something in the wrong place. 

In this case, I wasn't even trying to discuss anything, so I figured the the developers would just see the comment and go, "ok, we'll keep that in mind", then move on.

As for, https://github.com/photoprism/photoprism/issues/782#issuecomment-2029475112 I was not demanding anything. I was making a request. I did my best to use non-demanding language. Heck, I didn't even say anything like "do this" or "don't do this". I just said "I **hope** you don't". I do see how my last paragraph could give the comment the impression of making a demand, though. So I'm sorry about that.

I'm out of time, so I'll just say that I really appreciate the work open source developers do, but I also believe that figuring out how to deal with the bad "customers" is part of the job. Even though you are not being paid for it. That comment was the wrong way to deal with someone. Either respond politely, or don't respond at all. If any of the projects I've pushed out as open source gain any users, that is how I'm planning on dealing with them.

Anyway, I don't plan on continuing this discussion, but if I do, I'll post something in the forum instead of here.

If it is possible, maybe just break these comments off and move them into Discussions? Is that even possible?

**Comment by lastzero** [comment]
I'm happy to report that we have started to implement the remaining functionality for OpenID Connect (SSO) support, i.e. new config options and a guest role for our Community Edition. We expect it to be ready for testing in the coming weeks.

**Comment by lastzero** [comment]
Here is a preview of the PhotoPrism login page with OIDC enabled:

![oidc-login](https://github.com/photoprism/photoprism/assets/301686/58e89668-2404-4973-8f6a-e228be389e6c)

Note that you must have HTTPS/TLS enabled on your PhotoPrism instance and the OpenID Connect identity provider in order to authenticate against it. The newly added config options for this are as follows:

| Environment              | CLI Flag        | Default            | Description                                                                                              |
|--------------------------|-----------------|--------------------|----------------------------------------------------------------------------------------------------------|
| PHOTOPRISM_OIDC_URI      | --oidc-uri      |                    | identity provider issuer `URI` for single sign-on via OpenID Connect, e.g. "https://accounts.google.com" |
| PHOTOPRISM_OIDC_CLIENT   | --oidc-client   |                    | client `ID` for single sign-on via OpenID Connect                                                        |
| PHOTOPRISM_OIDC_SECRET   | --oidc-secret   |                    | client `SECRET` for single sign-on via OpenID Connect                                                    |
| PHOTOPRISM_OIDC_PROVIDER | --oidc-provider |                    | custom identity provider `NAME`, e.g. "Google"                                                           |
| PHOTOPRISM_OIDC_ICON     | --oidc-icon     |                    | custom identity provider icon `URI`                                                                      |
| PHOTOPRISM_OIDC_REDIRECT | --oidc-redirect |                    | automatically redirect unauthenticated users to the configured identity provider                         |
| PHOTOPRISM_OIDC_REGISTER | --oidc-register |                    | allow new users to create an account when they sign in with OpenID Connect                               |
| PHOTOPRISM_OIDC_USERNAME | --oidc-username | preferred_username | preferred username `CLAIM` for new OpenID Connect users (preferred_username, email, nickname)            |
| PHOTOPRISM_OIDC_WEBDAV   | --oidc-webdav   |                    | allow new OpenID Connect users to use WebDAV when they have a role that allows it                        |
| PHOTOPRISM_DISABLE_OIDC  | --disable-oidc  |                    | disable single sign-on via OpenID Connect, even if an identity provider has been configured              |

**Once the redirect back from the OIDC provider is working as well, we can test compatibility with other providers (Keycloak is used for development) and gather feedback from you to further improve the implementation. We also welcome early feedback on the names and descriptions of the config options shown above.**

**Comment by lastzero** [comment]
Our Developer Guide has been updated to include basic documentation of the OIDC authentication flow, service discovery, and related resources: https://docs.photoprism.app/developer-guide/api/oidc/

**Comment by lastzero** [comment]
This commit implements the `GET /api/v1/oidc/redirect` endpoint, so authentication against OIDC Identity Providers should generally work. Note that further refinement/testing is needed before we can release this, and that the `PHOTOPRISM_OIDC_REDIRECT` config option is not yet implemented, i.e. setting this option has no effect yet.

The [preview build on Docker Hub](https://hub.docker.com/r/photoprism/photoprism/tags?page=1&name=preview) has been updated so you can help test what is already implemented:
- https://docs.photoprism.app/getting-started/updates/#development-preview
- https://docs.photoprism.app/release-notes/#development-preview

**Comment by twsouthwick** [comment]
I gave this a test and it works great with Authelia. Thanks so much for this - much easier to add friends/family who are already in my SSO. A couple points of feedback:

- Users are automatically added to "guest". Are the granular roles going to continue to be under the membership versions? Either way, it would be nice to have a way to map OIDC roles/groups to a photoprism group with the auto-register. For now, the cli works, but this is a nice to have
-  PHOTOPRISM_OIDC_WEBDAV=true set the WebDAV to "Disabled" - I couldn't tell from the docs if it's expecting a boolean or something else

Overall, much appreciated functionality! One less password to remember.

**Comment by twsouthwick** [comment]
As a follow up suggestion, things like `PHOTOPRISM_OIDC_WEBDAV` would make a lot of sense to be `PHOTOPRISM_OIDC_WEBDAV_ROLE` or something similar where you supply the OIDC role so not all users are automatically allowed webdav access

**Comment by lastzero** [comment]
Thanks for testing it, @twsouthwick! That's very helpful.

> Users are automatically added to "guest". Are the granular roles going to continue to be under the membership versions?

We considered giving the ability to register higher privileged accounts, either through a config option or OIDC claims as you can see in this feature branch we used for testing:

https://github.com/photoprism/photoprism/blob/f9fda55d431a1c2e058dbc8c1807270e843c4867/internal/api/auth.go#L52-L58

However, there are some scenarios where this could lead to malicious or accidental privilege escalation, e.g. when inexperienced users use Google as Identity Provider and set the default role to "admin".

At this point, I don't believe the convenience gain for users having a custom identity provider (so they have full control over who uses it and who can sign up) is worth the risk - especially if it's just a handful of users where you need to change the role after they've created an account. Would you agree with that assessment?

> PHOTOPRISM_OIDC_WEBDAV=true set the WebDAV to "Disabled" - I couldn't tell from the docs if it's expecting a boolean or something else

For a list of accepted bool values, see (it accepts everything that makes sense e.g. 0, 1, true, false,...):
- https://pkg.go.dev/strconv#ParseBool

> As a follow up suggestion, things like PHOTOPRISM_OIDC_WEBDAV would make a lot of sense to be PHOTOPRISM_OIDC_WEBDAV_ROLE or something similar where you supply the OIDC role so not all users are automatically allowed webdav access

For security reasons, guest users cannot get WebDAV access or any other type of direct file system access. So, as mentioned in the usage description, this flag only has an effect if the user role allows it in the first place, e.g. when the account role is later changed to "user" or "admin" - see the "Roles and Permissions" overview in our User Guide:
- https://docs.photoprism.app/user-guide/users/roles/

Basically, the `PHOTOPRISM_OIDC_WEBDAV` flag is just there so we don't have to make that choice for everyone, as some might assume it's enabled and others might assume it's disabled by default. This way it's transparent and users can't blame us for not having known about this option, e.g. when they give someone "user" rights but don't want them to have webdav access as well.

**Comment by lastzero** [comment]
With today's changes, the [OIDC config options](https://docs.photoprism.app/developer-guide/api/oidc/#config-options) (as [shown above](https://github.com/photoprism/photoprism/issues/782#issuecomment-2195858656)) should be fully functional. Much of the profile information provided by OIDC should now also be set when you sign up, and updated the next time you sign in (unless you have e.g. manually changed your display name in PhotoPrism).

Unless there are essential features missing, I would suggest keeping this feature scope so we can shift our focus to testing and release this as soon as possible... any [help with that](https://docs.photoprism.app/release-notes/#development-preview) is much appreciated! :gem:

**Comment by twsouthwick** [comment]
Thanks for the response. I've signed up for the essential level to try out the different user roles now that I can easily onboard my family.

> However, there are some scenarios where this could lead to malicious or accidental privilege escalation, e.g. when inexperienced users use Google as Identity Provider and set the default role to "admin".

> At this point, I don't believe the convenience gain for users having a custom identity provider (so they have full control over who uses it and who can sign up) is worth the risk - especially if it's just a handful of users where you need to change the role after they've created an account. Would you agree with that assessment?

I can see the point, but I'd prefer to have it tied to a claim and ensure the docs are clear. However, I would actually propose a different pattern, I think, then what you have above, to allow users to define their own role mapping which would reduce the chances of this:

1. Define a set of variables for role mapping, i.e. `PHOTOPRISM_OIDC_ADMIN_ROLE`, `PHOTOPRISM_OIDC_USER_ROLE`, etc
2. Identify precedence order. Regarding your concerns about inexperienced users, here you might go with the most restrictive group they're in; that is, if they're in the role for both admin and user, go with user. I'd prefer least restrictive (i.e. admin), but that's your call
3. If none of the role mappings are defined, default to Guest

This way, there's not a "default" role being added, but will check the groups retrieved from OIDC and the chance of that inexperienced user falling into an unintended state of disclosure becomes much smaller, as they would have to set the env variable AND add that role to the user.

As a bonus, it would also be great if this were to reevaluate a user's role when signing in (or at least when a new token is submitted)

**Comment by lastzero** [comment]
@twsouthwick Thanks for you support! Based on your comments, I assume you are using a self-hosted identity provider? If so, which one? It would be good to learn a bit more about your specific environment to better understand the context and potential attack vectors, and then compare possible solutions in terms of complexity and security. You are welcome to email us for this (or open a GitHub Discussion) as the comments under this issue can easily get lost due to their sheer number.

**Comment by lastzero** [comment]
When a new user signs in with OpenID Connect, their preferred username may already be registered. In this case, a random 6-digit number is now appended to resolve the conflict - see the updated documentation for further details:
- https://docs.photoprism.app/getting-started/advanced/openid-connect/#preferred-username

**Comment by lastzero** [comment]
Today's tests with Google as Identity Provider were successful after implementing a few changes to generate the username from additional OIDC claims (when `preferred_username` is not set) and resolve username conflicts. Final testing for the upcoming release will start next week. Have a great weekend, everyone! 🤗

**Comment by lastzero** [comment]
Our [development preview](https://hub.docker.com/r/photoprism/photoprism/tags?name=preview) build has been updated to include all the latest changes and improvements - since there are currently no more [known issues](https://docs.photoprism.app/known-issues/#openid-connect-oidc), this is a [release candidate](https://docs.photoprism.app/release-notes/#development-preview)! :tada:

**Comment by lastzero** [comment]
The additional changes implemented yesterday allow existing users to keep their login password when changing the authentication provider from "default" or "local" to "oidc" (now also [**in combination with 2FA**](https://docs.photoprism.app/user-guide/users/2fa/) if enabled for the account).

Our documentation has been updated accordingly:
- https://docs.photoprism.app/getting-started/advanced/openid-connect/#existing-accounts
- https://docs.photoprism.app/known-issues/#openid-connect-oidc

Also note that admins cannot change the authentication provider of their own account through the Admin UI, so that they don't accidentally lock themselves out e.g. by setting it to "none".

Our own tests have been successful. However, we will give you some more time to [**test the preview build**](https://docs.photoprism.app/release-notes/#development-preview) before tagging the new release. **Many thanks to everyone who contributed, especially @moximoti who got things rolling and did much of the necessary work!** :rainbow:

**Comment by lastzero** [comment]
Since some users had problems with the configuration, we updated our documentation with additional notes on the Issuer URI and Redirect URL to be used:

- https://docs.photoprism.app/getting-started/advanced/openid-connect/#issuer-uri
- https://docs.photoprism.app/getting-started/advanced/openid-connect/#redirect-url
